Simplify Combined Z3 Preconditions in z3_visitor.py Using z3.simplify

IbrahimBilal2005 opened the pull request:
## Proposed Changes

Fixes #1062

This pull request updates set_function_def_z3_constraints in z3_visitor.py to simplify the combined Z3 preconditions using z3.simplify(z3.And(...)).
Previously, each precondition was stored individually. This change combines and simplifies them into a single Z3 expression, improving efficiency and logical clarity.

Also updates the README and CHANGELOG files to reflect this change.
...

<details>
<summary>Screenshots of your changes (if applicable)</summary>

</details>

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |          |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |    x      |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |    x      |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |   x       |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] I have updated the project Changelog (this is required for all changes).
- [x] If this is my first contribution, I have added myself to the [list of contributors](https://github.com/pyta-uoft/pyta/blob/master/README.md#contributors).

After opening your pull request:

- [x] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [x] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

_(Include any questions or comments you have regarding your changes.)_

IbrahimBilal2005 commented:
Hi all! I'm a first-time contributor and this PR adds support for simplifying combined Precondition expressions using z3.simplify, as discussed in Issue #1062.

I can’t formally request reviewers from a fork, so tagging here.

Please let me know if there are any issues with this PR, I'll be happy to make improvements :)

Thanks in advance!
@david-yz-liu

david-yz-liu commented on CHANGELOG.md:
```diff
@@ -249,6 +249,7 @@ Custom checkers:
 - Added support for the `!=` operator and replaced dictionary indexing with `.get` in `ExprWrapper`.
 - Refactored `Z3Visitor` to use `safe_infer()` instead of `inferred()` and added handling of `AstroidError`.
 - Add `negate` attribute to `CFGEdge`
+- Simplified combined Z3 preconditions in `set_function_def_z3_constraints` using `z3.simplify`
```
This entry should be added to the "unreleased" section

david-yz-liu commented on README.md:
```diff
@@ -120,3 +120,4 @@ Jasmine Wu,
 Raine Yang,
 Philippe Yu,
 Yi Cheng (Michael) Zhao
+Ibrahim Bilal
```
Preserve the alphabetical order of this list, and make sure to add a comma

IbrahimBilal2005 commented:
@david-yz-liu  Thank you for the review! I've addressed the feedback and updated the tests accordingly — they now pass locally. Also fixed the checklist formatting in the PR description. Let me know if there's anything else you'd like me to adjust.

IbrahimBilal2005 commented:
@david-yz-liu Thanks for the feedback! 

I’ve updated the tests in test_z3_visitor.py +  test_z3_constraints.py, and addressed your other comments. 

Let me know if anything else needs to be adjusted :)

david-yz-liu commented on tests/test_cfg/test_z3_constraints.py:
```diff
@@ -18,8 +18,12 @@ def func(x: int, y: int) -> int:
+    assert all(
+        edge.z3_constraints[0][0] == expected
+        for edge in cfg.get_edges()
+        if 0 in edge.z3_constraints and edge.z3_constraints[0]
```
Adding extra guards to these tests shouldn't be necessary. The code change did update the value of `edge.z3_constraints` but should not change the overall structure of the list, nor the number of edges.

In general you should avoid unnecessary like this, as they either add confusion about what the expected behaviour is, or (in the case of failing tests) can mask deeper issues with the code.

The same comment applies throughout; I would expect to see only changes to the expected value of the constraints by calling `z3.simplify`, but no other test changes.

david-yz-liu commented on tests/test_z3_visitor.py:
```diff
@@ -289,37 +289,51 @@ def string_step_length(x: str, y: str):
     ],
 ]
 
+
```
Revert this change

david-yz-liu reviewed (commented):
@IbrahimBilal2005 thanks for your update. I left some more comments to provide feedback on the test cases you added.

IbrahimBilal2005 commented:
@david-yz-liu  Thanks again for the feedback! I've updated the Z3Visitor to simplify each individual precondition separately using z3.simplify(expr) instead of combining them with And(...). This preserves the structure and count of the constraints.

All tests either remained unchanged or required only updates to the expected values to reflect the simplified forms. No structural test changes were needed (many are now unchanged from the master version), and all tests now pass.

Let me know if you'd prefer I revert to the previous And(...) based simplification and update the tests accordingly.

david-yz-liu reviewed (approved):
Great, thanks @IbrahimBilal2005 for your contribution!
